amaranth/ml/train.py

```diff
@@ -52,24 +57,69 @@ def main():
       low_calorie_threshold=amaranth.LOW_CALORIE_THRESHOLD,
       high_calorie_threshold=amaranth.HIGH_CALORIE_THRESHOLD)
 
+  # Normalize input strings
+  # Step 1: convert strings to lowercase
+  # Step 2: filter out characters present in DISH_NAME_FILTERS
+  # Step 3: re-combine characters back to a normal string
+  calorie_data['description'] = calorie_data['description'].apply(
+      lambda desc: desc.lower(),  # Step 1
+  ).apply(
+      lambda desc: [char for char in desc
+                    if char not in DISH_NAME_FILTERS],  # Step 2
+  ).apply(
+      ''.join,  #Step 3
+  )
+
   # Do some preprocessing and calculations for encoding
-  corpus = calorie_data['description'].str.replace(',', '').str.lower()
+  corpus = calorie_data['description']
   vocab_size = lib.num_unique_words(corpus)
   tokenized_corpus = corpus.map(lambda desc: desc.split(' '))
   max_corpus_length = lib.max_sequence_length(tokenized_corpus)
 
-  # Create TextVectorization layer
-  vectorize_layer = keras.layers.experimental.preprocessing.TextVectorization(
-      max_tokens=vocab_size,
-      output_sequence_length=max_corpus_length,
-  )
-
-  vectorize_layer.adapt(calorie_data['description'].to_numpy())
+  # Encode 'description' into new column 'input'
+  calorie_data['tokenized'] = calorie_data.apply(
+      lambda row: text.one_hot(row['description'], vocab_size), axis=1)
+
+  # Create tokenizer to encode input text
+  tokenizer = dict()  # Dictionary mapping each unique word to a unique int
+  tokenizer_cnt = dict()  # Counts appearances of each word in tokenizer
+  for _, row in calorie_data.iterrows():
+    for idx, word in enumerate(row['description'].split()):
+      tokenizer[word] = row['tokenized'][idx]
```

Comment:
Does this actually create a unique int per word? I'd imagine it repeats from 0-n per sentence. Also, there's a decent chance words actually repeat so you'd want to make sure that doesn't happen either?

Reply:
It does create a unique int per word. The `row['tokenized']` series is a list of ints created using [keras' one-hot-encoding function](https://www.tensorflow.org/api_docs/python/tf/keras/preprocessing/text/one_hot) before the loop starts, this loop just populates the dictionary with those values so it can be serialized to json!

Reply:
Ah okay, that sounds good. Also worth noting it seems like it generates ints = [1, n] which means you can always keep OOV=0

Reply:
Good catch, that's convenient!

Reply:
Yup. Just make sure about the uniqueness guarantee? The site does say "unicity non-guaranteed"

Reply:
You're totally right. When I came across that I was transitioning to the TextVectorization layer, so I thought it wouldn't matter after that. But since we're back to rolling our own vectorization code that's something I'll have to fix

Reply:
Fixed in https://github.com/googleinterns/amaranth/pull/35/commits/27f9a674db14f5a5caf457c7d3e06ac2481e69c9